salt/_modules/metalk8s_volumes.py

```diff
@@ -0,0 +1,360 @@
+    """Create the backing storage device for the given volume.
+
+    Args:
+        name (str): volume name
+
+    Returns:
+        None
+
+    CLI Example:
+
+    .. code-block:: bash
+
+        salt '*' metalk8s_volumes.create example-volume
+    """
+    return _get_volume(name).create()
+
+
+def is_initialized(name):
+    """Check if the backing storage device is initialized for the given volume.
+
+    Args:
+        path (str): path of the sparse file
+
+    Returns:
+        bool: True if the backing storage device is initialized, otherwise False
+
+    CLI Example:
+
+    .. code-block:: bash
+
+        salt '*' metalk8s_volumes.is_initialized example-volume
+    """
+    return _get_volume(name).is_initialized
+
+
+def initialize(name):
+    """Initialize the backing storage device of the given volume.
+
+    Args:
+        name (str): volume name
+
+    Returns:
+        str: path to the associated loop device
+
+    CLI Example:
+
+    .. code-block:: bash
+
+        salt '*' metalk8s_volumes.initialize example-volume
+    """
+    return _get_volume(name).initialize()
+
+
+def is_formatted(name):
+    """Check if the given volume is formatted.
+
+    Args:
+        name (str): volume name
+
+    Returns:
+        bool: True if the volume is already formatted, otherwise False
+
+    CLI Example:
+
+    .. code-block:: bash
+
+        salt '*' metalk8s_volumes.volume_is_formatted example-volume
+    """
+    return _get_volume(name).is_formatted
+
+
+def format(name):
+    """Format the given volume.
+
+    Args:
+        name (str): volume name
+
+    Returns:
+        None
+
+    CLI Example:
+
+    .. code-block:: bash
+
+        salt '*' metalk8s_volumes.format example-volume
+    """
+    _get_volume(name).format()
+
+
+# Volume {{{
+
+
+class Volume(object):
+    """Volume interface."""
+
+    __metaclass__ = abc.ABCMeta
+
+    def __init__(self, volume):
+        self._volume = volume
+
+    @abc.abstractproperty
+    def exists(self):
+        """Does the backing storage device exists?"""
+        return
+
+    @abc.abstractmethod
+    def create(self):
+        """Create the backing storage device."""
+        return
+
+    @abc.abstractproperty
+    def is_initialized(self):
+        """Check if the backing storage device is initialized."""
+        return
+
+    @abc.abstractmethod
+    def initialize(self):
+        """Initialize the backing storage device."""
+        return
+
+    @abc.abstractproperty
+    def block_device(self):
+        """Path to the backing block device."""
+        return
+
+    @property
+    def is_formatted(self):
+        """Check if the volume is already formatted."""
+        fs_type = self['spec.storageClassName.parameters.fsType']
+        return __salt__['disk.fstype'](self.block_device) == fs_type
+
+    def format(self):
+        """Format the volume.
+
+        The volume is formatted according to its StorageClass.
+        """
+        params = self['spec.storageClassName.parameters']
```

Comment:
I'd argue this needs a failsafe which checks whether the device is formatted already, and if so, errors out.

Otherwise, if someone creates a `rawBlockDevice` `Volume` for `/dev/sda1` by mistake, we could wipe a whole server (root FS at least). Not cool.

Reply:
Indeed & Done.